File: shared/ride-engine/ride-engine.js
export class RideEngine {
  constructor(config = {}) {
    this.config = {
      gravityMs2: 9.80665,
      positiveGThreshold: 1.0,
      airtimeThreshold: 0.3,
      gps: {
        maxHorizontalAccuracyM: 40,
        stationarySpeedMs: 0.8,
        minimumAcceptedMovementM: 1.5,
        maxImpliedSpeedMs: 90,
        accuracyMultiplier: 0.55,
      },
      ...config,
    };
    this.reset();
  }

  reset() {
    this.calibration = null;
    this.lastLocation = null;
    this.distanceM = 0;
    this.positiveGSum = 0;
    this.positiveGCount = 0;
    this.acceptedLocations = 0;
    this.rejectedLocations = 0;
  }

  setCalibration({ up, lateral, forward, source = 'manual' }) {
    this.calibration = { up, lateral, forward, source };
  }

  processMotion(sample) {
    const vector = [sample.x, sample.y, sample.z];
    const project = axis => axis ? vector.reduce((sum, value, index) => sum + value * axis[index], 0) : null;
    const normal = project(this.calibration?.up) ?? sample.z;
    const lateral = project(this.calibration?.lateral) ?? sample.x;
    const longitudinal = project(this.calibration?.forward) ?? sample.y;
    const total = Math.hypot(normal, lateral, longitudinal);

    if (normal > this.config.positiveGThreshold) {
      this.positiveGSum += normal;
      this.positiveGCount += 1;
    }

    return {
      ...sample,
      normal,
      lateral,
      longitudinal,
      total,
      positiveGAverage: this.positiveGCount ? this.positiveGSum / this.positiveGCount : null,
      airtime: normal < this.config.airtimeThreshold,
    };
  }

  processLocation(point) {
    const gps = this.config.gps;
    if (!Number.isFinite(point.accuracy) || point.accuracy > gps.maxHorizontalAccuracyM) {
      this.rejectedLocations += 1;
      return { accepted: false, reason: 'accuracy' };
    }

    if (!this.lastLocation) {
      this.lastLocation = point;
      this.acceptedLocations += 1;
      return { accepted: true, segmentDistanceM: 0, totalDistanceM: 0 };
    }

    const dt = Math.max(0.001, point.t - this.lastLocation.t);
    const distance = haversine(this.lastLocation, point);
    const impliedSpeed = distance / dt;
    const reportedSpeed = Number.isFinite(point.speed) ? Math.max(0, point.speed) : null;
    const uncertainty = Math.max(point.accuracy, this.lastLocation.accuracy) * gps.accuracyMultiplier;

    const stationary = reportedSpeed !== null && reportedSpeed < gps.stationarySpeedMs && distance <= uncertainty;
    const tooSmall = distance < gps.minimumAcceptedMovementM && dt < 2;
    const impossible = impliedSpeed > gps.maxImpliedSpeedMs;

    if (stationary || tooSmall || impossible) {
      this.rejectedLocations += 1;
      return { accepted: false, reason: stationary ? 'stationary' : tooSmall ? 'minimum-movement' : 'implied-speed' };
    }

    this.distanceM += distance;
    this.lastLocation = point;
    this.acceptedLocations += 1;
    return { accepted: true, segmentDistanceM: distance, totalDistanceM: this.distanceM };
  }

  summary() {
    return {
      distanceM: this.distanceM,
      positiveGAverage: this.positiveGCount ? this.positiveGSum / this.positiveGCount : null,
      positiveGSamples: this.positiveGCount,
      acceptedLocations: this.acceptedLocations,
      rejectedLocations: this.rejectedLocations,
    };
  }
}

function haversine(a, b) {
  const radius = 6371000;
  const radians = Math.PI / 180;
  const dLat = (b.lat - a.lat) * radians;
  const dLon = (b.lon - a.lon) * radians;
  const q = Math.sin(dLat / 2) ** 2 + Math.cos(a.lat * radians) * Math.cos(b.lat * radians) * Math.sin(dLon / 2) ** 2;
  return 2 * radius * Math.asin(Math.sqrt(q));
}

File: native-android/app/src/main/java/de/ridetracker/engine/RideEngine.kt
package de.ridetracker.engine

import android.location.Location
import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.sin
import kotlin.math.sqrt

data class Vector3(val x: Double, val y: Double, val z: Double)
data class SeatCalibration(val up: Vector3, val lateral: Vector3, val forward: Vector3, val source: String)
data class MotionInput(val t: Double, val x: Double, val y: Double, val z: Double)
data class MotionOutput(
    val t: Double,
    val normalG: Double,
    val lateralG: Double,
    val longitudinalG: Double,
    val totalG: Double,
    val positiveGAverage: Double?,
    val airtime: Boolean
)
data class LocationInput(
    val t: Double,
    val latitude: Double,
    val longitude: Double,
    val accuracy: Double,
    val speed: Double?
)
data class LocationResult(val accepted: Boolean, val reason: String? = null, val segmentDistanceM: Double = 0.0, val totalDistanceM: Double = 0.0)

class RideEngine {
    var calibration: SeatCalibration? = null
    var distanceM: Double = 0.0
        private set
    var acceptedLocations: Int = 0
        private set
    var rejectedLocations: Int = 0
        private set

    private var lastLocation: LocationInput? = null
    private var positiveGSum = 0.0
    private var positiveGCount = 0

    private val positiveGThreshold = 1.0
    private val airtimeThreshold = 0.3
    private val maxAccuracyM = 40.0
    private val stationarySpeedMs = 0.8
    private val minimumMovementM = 1.5
    private val maxImpliedSpeedMs = 90.0

    fun reset() {
        distanceM = 0.0
        acceptedLocations = 0
        rejectedLocations = 0
        lastLocation = null
        positiveGSum = 0.0
        positiveGCount = 0
    }

    fun processMotion(input: MotionInput): MotionOutput {
        val vector = Vector3(input.x, input.y, input.z)
        val normal = calibration?.let { dot(vector, it.up) } ?: input.z
        val lateral = calibration?.let { dot(vector, it.lateral) } ?: input.x
        val longitudinal = calibration?.let { dot(vector, it.forward) } ?: input.y
        val total = sqrt(normal * normal + lateral * lateral + longitudinal * longitudinal)

        if (normal > positiveGThreshold) {
            positiveGSum += normal
            positiveGCount += 1
        }

        return MotionOutput(
            t = input.t,
            normalG = normal,
            lateralG = lateral,
            longitudinalG = longitudinal,
            totalG = total,
            positiveGAverage = if (positiveGCount > 0) positiveGSum / positiveGCount else null,
            airtime = normal < airtimeThreshold
        )
    }

    fun processLocation(point: LocationInput): LocationResult {
        if (!point.accuracy.isFinite() || point.accuracy > maxAccuracyM) {
            rejectedLocations += 1
            return LocationResult(false, "accuracy")
        }
        val previous = lastLocation
        if (previous == null) {
            lastLocation = point
            acceptedLocations += 1
            return LocationResult(true)
        }

        val dt = max(0.001, point.t - previous.t)
        val distance = haversine(previous, point)
        val impliedSpeed = distance / dt
        val reportedSpeed = point.speed?.coerceAtLeast(0.0)
        val uncertainty = max(point.accuracy, previous.accuracy) * 0.55
        val stationary = reportedSpeed != null && reportedSpeed < stationarySpeedMs && distance <= uncertainty
        val tooSmall = distance < minimumMovementM && dt < 2.0
        val impossible = impliedSpeed > maxImpliedSpeedMs

        if (stationary || tooSmall || impossible) {
            rejectedLocations += 1
            return LocationResult(false, if (stationary) "stationary" else if (tooSmall) "minimum-movement" else "implied-speed")
        }

        distanceM += distance
        lastLocation = point
        acceptedLocations += 1
        return LocationResult(true, segmentDistanceM = distance, totalDistanceM = distanceM)
    }

    private fun dot(a: Vector3, b: Vector3) = a.x * b.x + a.y * b.y + a.z * b.z

    private fun haversine(a: LocationInput, b: LocationInput): Double {
        val result = FloatArray(1)
        Location.distanceBetween(a.latitude, a.longitude, b.latitude, b.longitude, result)
        return result[0].toDouble()
    }
}
